File: python/mysql.py
import mysql as sql

class mysql():
    def __init__(self, databaseConnect, debug=False):
        """
        Constructor method for the `mysql` class.

        Parameters:
        - databaseConnect: MySQL database connection object.
        - debug (optional): If True, enables debug mode.

        Attributes:
        - database: MySQL database connection.
        - results: Results of the last query.
        - debug: Debug mode status.
        - isLocked: Flag indicating whether the instance is currently locked.
        - Normalcursor: Placeholder for a normal cursor.
        """
        self.database = databaseConnect
        self.results = None
        self.debug = debug
        self.isLocked = False
        self.Normalcursor = False

    def start(self):
        """
        Initiates a transaction on the connected database.

        Usage:
        db = mysql(databaseConnect=your_database_connection)
        db.start()
        """
        self.database.start_transaction()

    def commit(self):
        """
        Commits the current transaction to the database.

        Usage:
        db.commit()
        """
        self.database.commit()

    def query(self, query: str, params: list, column=True):
        """
        Executes a SQL query on the database.

        Parameters:
        - query: SQL query string.
        - params: List of parameters for the query.
        - column (optional): If True, returns results as a list of dictionaries.

        Returns:
        - Boolean indicating success or failure.

        Usage:
        result = db.query("SELECT * FROM your_table WHERE your_condition", params=[...])
        if result:
            print("Query successful. Result:", db.result())
        else:
            print("Query failed. Error:", db.error())
        db.commit()
        """
        if not self.isLocked:
            self.database.commit()
        err = 0
        result = ''
        try:
            if self.isLocked:
                mycursor = self.cursor
            else:
                if self.Normalcursor:
                    mycursor = self.Normalcursor
                else:
                    mycursor = self.database.cursor()
                    self.Normalcursor = mycursor
            mycursor.execute(query, params)
            if column:
                columns = mycursor.description
                result = [{columns[index][0]: column for index, column in enumerate(value)} for value in mycursor.fetchall()]
            else:
                result = mycursor.fetchall()
            returndata = {
                "result": result,
                "length": len(result),
                "error": int(err),
                "affectedRows": mycursor.rowcount
            }
            self.database.commit()
            self.results = returndata
            return True
        except sql.connector.Error as err:
            err = err
            returndata = {
                "result": result,
                "length": len(result),
                "error": int(err.errno),
                "affectedRows": mycursor.rowcount
            }
            if self.debug:
                print(mycursor.statement)
            if self.isLocked:
                self.database.rollback()  # Rollback the transaction in case of an error
                self.isLocked = False
            self.results = returndata
            return False

    def result(self):
        """
        Returns the result of the last executed query.

        Usage:
        result = db.result()
        """
        if self.results['result']:
            return self.results['result']
        # just returns the result.
        return False

    def error(self):
        """
        Returns the error code of the last executed query.

        Usage:
        error_code = db.error()
        """
        if int(self.results['error']) != 0:
            return int(self.results['error'])
        return False

    def changedRows(self):
        """
        Returns the number of affected rows by the last executed query.

        Usage:
        affected_rows = db.changedRows()
        """
        return self.results['affectedRows']

    def length(self):
        """
        Returns the length of the result set from the last executed query.

        Usage:
        result_length = db.length()
        """
        return self.results['length']

    def lockingQuery(self, query: str, params: list, column=True, lock_type=None):
        """
        Executes a SQL query with a specified lock type.

        Parameters:
        - query: SQL query string.
        - params: List of parameters for the query.
        - column (optional): If True, returns results as a list of dictionaries.
        - lock_type (optional): Lock type, either "FOR UPDATE" or "FOR DELETE".

        Returns:
        - Boolean indicating success or failure.

        Usage:
        db.lockingQuery("SELECT * FROM your_table WHERE your_condition", params, lock_type="FOR UPDATE")
        """
        err = 0
        result = ''

        if lock_type and lock_type not in ("FOR UPDATE", "FOR DELETE"):
            raise ValueError("Lock type must be 'FOR UPDATE' or 'FOR DELETE'")
        self.isLocked = True
        try:
            self.database.commit()
            # get new data

            mycursor = self.database.cursor()
            mycursor.execute("START TRANSACTION")
            mycursor.execute("SET SESSION TRANSACTION ISOLATION LEVEL SERIALIZABLE")
            if lock_type:
                mycursor.execute(query + f" {lock_type}", params)  # Apply the specified lock
            else:
                # don't use this this.
                raise NotImplementedError()
                mycursor.execute(query, params)

            if column:
                columns = mycursor.description
                result = [{columns[index][0]: column for index, column in enumerate(value)} for value in mycursor.fetchall()]
            else:
                result = mycursor.fetchall()

            returndata = {
                "result": result,
                "length": len(result),
                "error": int(err),
                "affectedRows": mycursor.rowcount
            }

            self.results = returndata
            self.cursor = mycursor
            return True
        except sql.connector.Error as err:
            returndata = {
                "result": result,
                "length": len(result),
                "error": int(err.errno),
                "affectedRows": mycursor.rowcount
            }

            if self.debug:
                print(mycursor.statement)

            self.results = returndata
            self.database.rollback()  # Rollback the transaction in case of an error
            self.isLocked = False
            return False

File: go/mysql.go
package mysqlconnector

import (
	"database/sql"
	"fmt"

	_ "github.com/go-sql-driver/mysql"
)

// MySQL struct represents our MySQL database connection handler.
type MySQL struct {
	db     *sql.DB
	result map[string]interface{}
	debug  bool
}

// NewMySQL initializes a new MySQL instance.
func NewMySQL(db *sql.DB, debug bool) *MySQL {
	return &MySQL{
		db:     db,
		result: make(map[string]interface{}),
		debug:  debug,
	}
}

// Start begins a new transaction.
func (m *MySQL) Start() (*sql.Tx, error) {
	tx, err := m.db.Begin()
	if err != nil {
		return nil, err
	}
	return tx, nil
}

// Commit commits the current transaction.
func (m *MySQL) Commit(tx *sql.Tx) error {
	return tx.Commit()
}

// Rollback rolls back the current transaction.
func (m *MySQL) Rollback(tx *sql.Tx) error {
	return tx.Rollback()
}

// Query executes a SQL query within a transaction
func (m *MySQL) Query(tx *sql.Tx, query string, params ...interface{}) error {
	rows, err := tx.Query(query, params...)
	if err != nil {
		return err
	}
	defer rows.Close()

	columns, err := rows.Columns()
	if err != nil {
		return err
	}

	var result []map[string]interface{}
	for rows.Next() {
		row := make(map[string]interface{})
		values := make([]interface{}, len(columns))
		valuePtrs := make([]interface{}, len(columns))
		for i := range columns {
			valuePtrs[i] = &values[i]
		}
		if err := rows.Scan(valuePtrs...); err != nil {
			return err
		}
		for i, col := range columns {
			val := values[i]
			switch v := val.(type) {
			case []byte:
				row[col] = string(v)
			default:
				row[col] = v
			}
		}
		result = append(result, row)
	}

	m.result["result"] = result
	m.result["length"] = len(result)

	return nil
}

// Result returns the result of the last executed query.
func (m *MySQL) Result() []map[string]interface{} {
	if result, ok := m.result["result"].([]map[string]interface{}); ok {
		return result
	}
	return nil
}

func main() {
	// Connect to MySQL database
	db, err := sql.Open("mysql", "username:password@tcp(localhost:3306)/database_name")
	if err != nil {
		fmt.Println("Error connecting to database:", err)
		return
	}
	defer db.Close()

	mysqlInstance := NewMySQL(db, true)

	// Example usage: Perform a transaction with a query
	tx, err := mysqlInstance.Start()
	if err != nil {
		fmt.Println("Error starting transaction:", err)
		return
	}
	defer func() {
		if err != nil {
			mysqlInstance.Rollback(tx)
		} else {
			err = mysqlInstance.Commit(tx)
			if err != nil {
				fmt.Println("Error committing transaction:", err)
			}
		}
	}()

	query := "SELECT * FROM your_table WHERE your_condition"
	params := []interface{}{} // Provide query parameters if needed
	err = mysqlInstance.Query(tx, query, params...)
	if err != nil {
		fmt.Println("Query execution error:", err)
		return
	}

	results := mysqlInstance.Result()
	fmt.Println("Query successful. Result:", results)
}
